### PR TITLE
feat(claims): add runtime-scoped claim reads

### PR DESCRIPTION
**Pull request by jonathanhaaswriter**
## Summary
- add runtime-scoped claim listing to BootstrapService plus the mirrored HTTP surface
- extend claim storage and services with filtered reads over persisted claim batches
- add SDK helpers for reading claims from TypeScript and Python integrations

## Testing
- make verify
- python3 -m py_compile sdk/python/cerebro_sdk/client.py sdk/python/cerebro_sdk/__init__.py
- cd sdk/typescript && npx -y -p typescript tsc -p tsconfig.json

### CODE REVIEW

**Comment by jonathanhaaswriter** [comment]
@droid Please perform a deep code review and security review of this PR. Leave inline comments only for high-confidence, actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed conflict-resolution fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Added service-layer nil timestamp preservation and regression coverage; make verify passed locally.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
### Droid sweep status

Latest sweep: `2026-04-29 22:09 UTC`.

- Unresolved non-outdated review threads: `0`
- CI `verify`: `SUCCESS`
- `droid-review`: `not present`

I’ll keep watching for new review findings/check failures and address them as they appear.

**Comment by jonathanhaaswriter** [comment]
Re-reviewed the still-open outdated thread(s) on the latest branch tip. The branch already URL-encodes `runtime_id` as a single path segment in the Python client and caps `ListClaims`, so no code change was needed. Validation: `go test ./internal/claims -run TestListClaimsAppliesDefaultLimit -count=1`. Remaining assigned unresolved threads after this audit: none.

**Comment by jonathanhaaswriter** [comment]
Droid sweep audit complete (final pass)

- Re-scanned all live review threads, including outdated ones
- Verified each remaining thread against the current branch tip
- Resolved every actionable thread (or fixed it inline and pushed)
- 0 unresolved review threads across the whole open-PR stack at this time

PR #494 (fix/feedback-main-followup) and PR #495 (feat/droid-creation-workflows) received additional fixes for SSRF (DNS-rebind) and workflow branch-name validation in this pass.

**Comment by jonathanhaaswriter** [comment]
@droid review

**Comment by jonathanhaaswriter** [comment]
@droid please harden this PR end-to-end:

1. Re-review the diff for residual bugs, races, edge cases, security gaps, and untested paths.
2. Apply the smallest correct fixes for any high-confidence issues you find.
3. Strengthen test coverage: add regression tests for every prior review finding on this PR plus the new behavior, including failure paths, boundary inputs, concurrency, and not-found / typed-nil interface guards.
4. Run focused tests and `make verify`. If validators fail, fix until green.
5. Commit and push the fixes to this PR's branch (or, if you lack push permission in this workflow, leave concrete inline suggestions and a summary comment).

Do not change the PR scope, do not modify the default branch, and do not merge.

**Comment by jonathanhaaswriter** [comment]
@droid review

File: internal/statestore/postgres/claims.go [hunk 1]
```diff
@@ -219,6 +290,52 @@ func cloneClaimEntityRef(ref *cerebrov1.EntityRef) *cerebrov1.EntityRef {
 	}
 }
 
+func claimRecordFromJSON(runtimeID string, tenantID string, payload string) (*ports.ClaimRecord, error) {
+	if strings.TrimSpace(payload) == "" {
+		return nil, errors.New("claim json is required")
+	}
+	claim := &cerebrov1.Claim{}
+	if err := protojson.Unmarshal([]byte(payload), claim); err != nil {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/statestore/postgres/claims.go]: Fixed in 9152e8e2a by decoding persisted claim JSON with `protojson.UnmarshalOptions{DiscardUnknown: true}` and adding a forward-compatibility regression.

**Comment by jonathanhaaswriter** [comment]
@droid review

Please re-review the runtime-scoped claim reads deeply. Read APIs over claim data have direct exposure consequences:

1. **Tenant/runtime scoping**: confirm every claim list path (BootstrapService, HTTP, SDKs) filters by runtime+tenant before any data leaves the store, and that filter parameters cannot be coerced into a cross-tenant read via missing/empty/wildcard values.
2. **Authorization and visibility**: verify callers can only see claims they’re authorized for, that source-event filtering does not allow probing for claims belonging to events outside the caller’s scope, and that error messages don’t reveal whether out-of-scope claims exist.
3. **Pagination, ordering, and limits**: re-check that listing paths enforce sane page size caps, deterministic ordering, stable cursors, and bounded total results so a caller cannot pull entire claim history in one request.
4. **Filter validation**: ensure all filters (runtime id, source-event id, entity refs, time ranges) are parsed, validated, and rejected as `InvalidArgument` when malformed, with no SQL injection or unparameterized fragments.
5. **SDK parity**: verify Python and TypeScript readers produce identical results for the same filter inputs, including edge cases (no claims, very large pages, missing optional filters).
6. **Performance and indexing**: check that the new reads use the right indexes, do not require scans on large claim batches, and degrade gracefully under load (timeouts, cancellation propagation).
7. **Forward-compat**: confirm the read contract is stable for PR 474 (snapshot reconciliation), 471 (jira posture findings), and the SDK examples that rely on filter semantics.

**Comment by jonathanhaaswriter** [comment]
Closing this PR — its substantive feature work is already in `main` via the earlier consolidation rollups, and the unresolved factory-droid findings posted on this PR have been addressed in main via the follow-up PRs:

- #497 (`fix(github): drop non-cursor opaque from terminal audit checkpoints`)
- #499 (`fix: address post-merge feedback across pending stacks`)

PR #499 specifically addresses the findings posted on this stack:

- Reject malformed `root_urn` values in `GetEntityNeighborhood` (PR #460)
- Map `refuted` and `superseded` claim statuses to projection deletion (PR #463)
- Tenant-scope `ListClaims` reads (PR #464)
- Per-runtime serialization of `WriteClaims` (PR #474)
- Restore null-array normalization in `graph_summary` (PR #470)
- Surface blank-root errors from `graphLayering` (PR #468)

This branch has diverged ~50k lines from `main` and rebasing it cleanly is no longer practical; closing here keeps the open queue accurate. If anything from this stack still needs to land, please open a fresh PR cut from `main`.
